src/server/network/network.ts

```diff
@@ -0,0 +1,105 @@
+import dgram = require('dgram');
+
+import { WebSocketServer } from './web_socket_server';
+import { WebSocket } from './web_socket_server';
+
+type Opts = {
+  teamAAddress: string;
+  teamBAddress: string;
+};
+
+/**
+ * The server component of a UDP proxy running over web sockets. Acts as a gateway to the UDP network.
+ * All clients currently share a single UDP connection, mostly for performance reasons. Could potentially be
+ * improved to have more intelligent multiplexing.
+ */
+export class WebSocketProxyUDPServer {
+  constructor(
+    private server: WebSocketServer,
+    private teamAAddress: string,
+    private teamBAddress: string,
+  ) {
+    server.onConnection(this.onClientConnection);
+    this.teamAAddress = teamAAddress;
+    this.teamBAddress = teamBAddress;
+  }
+
+  static of(
+    server: WebSocketServer,
+    { teamAAddress, teamBAddress }: Opts,
+  ): WebSocketProxyUDPServer {
+    return new WebSocketProxyUDPServer(server, teamAAddress, teamBAddress);
+  }
+
+  private onClientConnection = (socket: WebSocket) => {
+    WebSocketServerClient.of(socket, this.teamAAddress, this.teamBAddress);
+  };
+}
+
+class WebSocketServerClient {
+  private connected: boolean;
+  private offListenMap: Map<string, () => void>;
+  private udpServer: dgram.Socket;
+
+  constructor(
+    private socket: WebSocket,
+    private teamAAddress: string,
+    private teamBAddress: string,
+  ) {
+    this.connected = false;
+    this.offListenMap = new Map();
+    this.teamAAddress = teamAAddress;
+    this.teamBAddress = teamBAddress;
+
+    // Create a UDP4 socket. Reuse the address in the event another process is already using it
+    this.udpServer = dgram.createSocket({ type: 'udp4', reuseAddr: true });
```

Comment:
This means each browser client opens a new UDP socket.

IIUC we could/should just share the same socket among all clients?

Reply:
How do we do that?